test/integration/tcp_proxy_integration_test.cc

```diff
@@ -483,9 +489,97 @@ TEST_P(TcpProxyIntegrationTest, AccessLog) {
   // Test that all three addresses were populated correctly. Only check the first line of
   // log output for simplicity.
   EXPECT_THAT(log_result,
-              MatchesRegex(fmt::format(
-                  "upstreamlocal={0} upstreamhost={0} downstream={1} sent=5 received=0\r?\n.*",
-                  ip_port_regex, ip_regex)));
+              MatchesRegex(fmt::format("upstreamlocal={0} upstreamhost={0} downstream={1} "
+                                       "sent=5 received=0 "
+                                       "DOWNSTREAM_WIRE_BYTES_SENT=5 "
+                                       "DOWNSTREAM_WIRE_BYTES_RECEIVED=0 "
+                                       "UPSTREAM_WIRE_BYTES_SENT=0 "
+                                       "UPSTREAM_WIRE_BYTES_RECEIVED=5"
+                                       "\r?.*",
+                                       ip_port_regex, ip_regex)));
+}
+
+// Make sure no bytes are logged when no data is sent.
+TEST_P(TcpProxyIntegrationTest, TcpProxyNoDataBytesMeter) {
+  setupByteMeterAccessLog();
+  initialize();
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  auto tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection->close());
+
+  test_server_.reset();
+  auto log_result = waitForAccessLog(listener_access_log_name_);
+  EXPECT_THAT(log_result, MatchesRegex(fmt::format("DOWNSTREAM_WIRE_BYTES_SENT=0 "
+                                                   "DOWNSTREAM_WIRE_BYTES_RECEIVED=0 "
+                                                   "UPSTREAM_WIRE_BYTES_SENT=0 "
+                                                   "UPSTREAM_WIRE_BYTES_RECEIVED=0"
+                                                   "\r?.*")));
+}
+
+// Test Byte Metering across multiple upstream/downstream connections
+TEST_P(TcpProxyIntegrationTest, TcpProxyBidirectionalBytesMeter) {
+  setupByteMeterAccessLog();
+  const int client_count = 3;
+  const int upstream_count = 3;
+
+  enableHalfClose(false);
+  config_helper_.addConfigModifier([this](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* cluster = bootstrap.mutable_static_resources()->mutable_clusters(0);
+    auto* load_assignment = cluster->mutable_load_assignment();
+    load_assignment->clear_endpoints();
+    for (int i = 0; i < upstream_count; ++i) {
+      auto locality = load_assignment->add_endpoints();
+      locality->add_lb_endpoints()->mutable_endpoint()->MergeFrom(
+          ConfigHelper::buildEndpoint(Network::Test::getLoopbackAddressString(version_)));
+    }
+  });
+
+  setUpstreamCount(upstream_count);
+  initialize();
+
+  std::vector<IntegrationTcpClientPtr> clients{client_count};
+  std::vector<FakeRawConnectionPtr> fake_connections{upstream_count};
+  auto indicies = std::vector<uint64_t>(upstream_count);
```

Comment:
nit: could do `std::vector<uint64_t> indicies{0,1,2}` to make it even more transparent what the values are + be consistent with the vectors above